Tools: Add eslint-excludelist (and use in CI)

anomiex opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
Like the existing phpcs-excludelist.json, this controls which files are allowed to have existing lint errors and which aren't.

It's now used in bin/eslint-excludelist.json.

This also renames the existing "PHP linting" workflow to just "Linting", adds jobs to run eslint and eslint-changed to it, and removes the existing misnamed "JavaScript & CSS lint" job from the "JS tests" workflow.
    
To merge this will need the "JavaScript & CSS lint" job un-required, and then after merging we can make the new ESLint jobs required.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None.

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Does CI pass?
* Try to make a commit introducing some JS lint errors. Does the pre-commit hook catch them?
* Try pushing such a PR. Does CI catch them?

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
* None needed.

brbrr reviewed (commented):
Something did not work well in my tests: https://github.com/Automattic/jetpack/pull/18180

For some reason, `eslint` has not been triggered for my changes, and eslint-changed have not flagged any errors

anomiex commented:
It works for me when I try reproducing your changes,

<details>
<summary>Transcript</summary>

```
$ git checkout add/eslint-excludelist
Previous HEAD position was fbc4f320d try this
Switched to branch 'add/eslint-excludelist'
Your branch is up to date with 'origin/add/eslint-excludelist'.
husky > post-checkout (node v12.19.0)
$ git checkout -b tmp
Switched to a new branch 'tmp'
husky > post-checkout (node v12.19.0)
$ patch -p1 < /tmp/0001-test-pre-commit-hook.patch
patching file _inc/client/lib/analytics/index.js
$ git commit -am test
husky > pre-commit (node v12.19.0)
Prettier formatting staged file: _inc/client/lib/analytics/index.js
$ tools/eslint-changed.js --ext .js,.jsx --git _inc/client/lib/analytics/index.js

/usr/local/src/automattic/jetpack/_inc/client/lib/analytics/index.js
  27:1  warning  Missing JSDoc block description           jsdoc/require-description
  27:1  warning  Missing JSDoc @returns declaration        jsdoc/require-returns
  28:0  warning  Missing JSDoc @param "group" description  jsdoc/require-param-description
  28:0  warning  Missing JSDoc @param "group" type         jsdoc/require-param-type
  29:0  warning  Missing JSDoc @param "name" description   jsdoc/require-param-description
  29:0  warning  Missing JSDoc @param "name" type          jsdoc/require-param-type
  47:1  warning  Missing JSDoc block description           jsdoc/require-description
  47:1  warning  Missing JSDoc @returns declaration        jsdoc/require-returns
  48:0  warning  Missing JSDoc @param "group" description  jsdoc/require-param-description
  48:0  warning  Missing JSDoc @param "group" type         jsdoc/require-param-type
  49:0  warning  Missing JSDoc @param "name" description   jsdoc/require-param-description
  49:0  warning  Missing JSDoc @param "name" type          jsdoc/require-param-type

✖ 12 problems (0 errors, 12 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
COMMIT ABORTED: The linter reported some problems. If you are aware of them and it is OK, repeat the commit command with --no-verify to avoid this check.
But please don't. Code is poetry.


husky > pre-commit hook failed (add --no-verify to bypass)
$ git commit -am test --no-verify
husky > prepare-commit-msg (node v12.19.0)
WARNING: git pre-commit hook was skipped!
[tmp dc1d227b0] [not verified] test
 1 file changed, 8 insertions(+)
$ patch -p1 < /tmp/0002-try-one-more-time.patch 
patching file _inc/client/lib/analytics/index.js
Hunk #1 succeeded at 11 with fuzz 1.
$ git commit -am test
husky > pre-commit (node v12.19.0)
Prettier formatting staged file: _inc/client/lib/analytics/index.js
$ tools/eslint-changed.js --ext .js,.jsx --git _inc/client/lib/analytics/index.js

/usr/local/src/automattic/jetpack/_inc/client/lib/analytics/index.js
  14:1  error    Unexpected var, use let or const instead  no-var
  14:5  warning  'testVar' is defined but never used       no-unused-vars

✖ 2 problems (1 error, 1 warning)
  1 error and 0 warnings potentially fixable with the `--fix` option.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
COMMIT ABORTED: The linter reported some problems. If you are aware of them and it is OK, repeat the commit command with --no-verify to avoid this check.
But please don't. Code is poetry.


husky > pre-commit hook failed (add --no-verify to bypass)
$ git commit -am test --no-verify
husky > prepare-commit-msg (node v12.19.0)
WARNING: git pre-commit hook was skipped!
[tmp 9e1b5e8c8] [not verified] test
 1 file changed, 1 insertion(+)
$ patch -p1 < /tmp/0003-try-this.patch 
patching file _inc/client/lib/analytics/index.js
Hunk #1 succeeded at 11 with fuzz 1.
$ git commit -am test
husky > pre-commit (node v12.19.0)
Prettier formatting staged file: _inc/client/lib/analytics/index.js
$ tools/eslint-changed.js --ext .js,.jsx --git _inc/client/lib/analytics/index.js

/usr/local/src/automattic/jetpack/_inc/client/lib/analytics/index.js
  15:1  error  'asfafa' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
COMMIT ABORTED: The linter reported some problems. If you are aware of them and it is OK, repeat the commit command with --no-verify to avoid this check.
But please don't. Code is poetry.


husky > pre-commit hook failed (add --no-verify to bypass)
```
</details>

And I see failures reported on your PR, both at https://github.com/Automattic/jetpack/pull/18180/checks?check_run_id=1600969737#step:7:10 and inline in https://github.com/Automattic/jetpack/pull/18180/files.

brbrr reviewed (approved):
Left a minor comment, but aside from that, it looks great!

While we at it, I'd like to mention that the existing linting action becomes even more complex. I wonder if it would make sense to add some docs explaining what's going on there in case someone would need to do a change later on.

anomiex commented on package.json:
```diff
@@ -65,9 +65,10 @@
+		"lint": "yarn lint-file .",
 		"lint-changed": "tools/eslint-changed.js --ext .js,.jsx --git",
 		"lint-file": "eslint --ext .js,.jsx",
+		"lint-required": "node -e \"const fs = require('fs'); fs.copyFileSync('.eslintignore','.eslintignore-required'); const w=fs.createWriteStream('.eslintignore-required',{flags:'a'}); w.write('\\n# bin/eslint-excludelist.json\\n'); w.end(JSON.parse(fs.readFileSync('bin/eslint-excludelist.json','utf8')).join('\\n')+'\\n')\" && yarn lint --ignore-path .eslintignore-required",
```
Someone might run it locally if they want to locally reproduce the CI job, or if they want to run with `--fix` over the files that are supposed to be clean.

No opinion on putting it into a file. Do we clutter package.json with the code to create the `.eslintignore-required' file, or clutter tools/ with the small script to do it?

brbrr replied:
🤷 Yeah, I don't know what would be the best here. Although, I just have a hard time parsing minimized js code 🙃

anomiex commented:
Pushed a PR to add some doc comments (and clean up a few other things). Please re-review.

brbrr reviewed (approved):
Looks good to me!
